includes/sanitizers/class-amp-script-sanitizer.php

```diff
@@ -69,4 +216,99 @@ public function sanitize() {
 			$this->did_convert_elements = true;
 		}
 	}
+
+	/**
+	 * Sanitize JavaScript script elements.
+	 *
+	 * This runs explicitly in the script sanitizer before the final validating sanitizer (tag-and-attribute) so that
+	 * the style sanitizer will be able to know whether there are custom scripts in the page, and thus whether tree
+	 * shaking can be performed.
+	 *
+	 * @since 2.2
+	 */
+	protected function sanitize_script_elements() {
+		$scripts = $this->dom->xpath->query( '//script[ not( @type ) or not( contains( @type, "json" ) ) ]' );
```

Comment:
Any reason for whitelisting `type=json` scripts? What about scripts containing HTML templates, for example?

Reply:
The reason for skipping JSON scripts is that they have no logic that can mutate the page, which is the main concern for script sanitization. If there is JSON on the page, it's not going to cause issues with tree shaking, for example. Nor will it negatively impact PX. So this is specifically sanitizing _JS_ scripts (which the method used to be called, and it could get renamed back).

Reply:
Renamed method and arg in dd49403704ddf71b4d9ae71f7e4d01357aeaf815 and 3070b837ae0cf7253eba1d8b5f3a6a789485274b.

Reply:
Got it.

> So this is specifically sanitizing _JS_ scripts (which the method used to be called, and it could get renamed back).

[According to MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-type), if the `type` attribute has any other value than `module` or a permitted JavaScript MIME type, the script will be treated as a data block, so I'm wondering if skipping only JSON scripts would be a shortsighted approach.

Reply:
Oh right. I forgot about examples like `<script type="text/plain">`

Reply:
Addressed in 03fb8b7152f035a61c2a43d6cecd85c11e0996d0